File: src/hey/RegularExpressionMatching.py
# problem: Regular Expression Matching
# author: everpuck
# date: 2/16/2019


"""
题目难度是hard，以下是官方参考答案之一，虽然速度很慢- -！
"""


class Solution:
    def isMatch(self, s: 'str', p: 'str') -> 'bool':
        if not p:
            return not s

        first_match = bool(s) and p[0] in {s[0], '.'}
        if len(p) >= 2 and p[1] == '*':
            return (self.isMatch(s, p[2:]) or
                    first_match and self.isMatch(s[1:], p))
        else:
            return first_match and self.isMatch(s[1:], p[1:])


if __name__ == '__main__':
    s = "miss issippi"
    p = "mis*is*p*."
    s = "ab"
    p = ".*"
    s = "a"
    p = ".*..a*"

    s = "ab"
    p = ".*..c*"
    s = 'aa'
    p = 'a'
    s = 'ab'
    p = '.*'

    solution = Solution()
    print(solution.isMatch(s, p))

File: src/hey/RegularExpressionMatching.java
/*
    problem: Regular Expression Matching
    author: everpuck
    date: 2/19/2019
*/


class Solution {
    // official answer
    public boolean isMatch(String s, String p) {
        if (p.isEmpty()) {
            return s.isEmpty();
        }
        boolean first_match = (!s.isEmpty() &&
                               (p.charAt(0) == s.charAt(0) || p.charAt(0) == '.'));
        if (p.length() >= 2 && p.charAt(1) == '*'){
            return (isMatch(s, p.substring(2)) ||
                    (first_match && isMatch(s.substring(1), p)));
        } else {
            return first_match && isMatch(s.substring(1), p.substring(1));
        }
    }
}


public class RegularExpressionMatching {
    public static void main(String[] args) {
        String s = "miss issippi";
        String p = "mis*is*p*.";
        Solution solution = new Solution();
        System.out.println(solution.isMatch(s, p));
    }
}
